app/services/framework_responses/bulk_updater.rb

```diff
@@ -0,0 +1,63 @@
+# frozen_string_literal: true
+
+module FrameworkResponses
+  class BulkUpdater
+    attr_accessor :person_escort_record, :response_values_hash, :errors
+
+    def initialize(person_escort_record, response_values_hash)
+      self.person_escort_record = person_escort_record
+      self.response_values_hash = response_values_hash
+      self.errors = {}
+    end
+
+    def call
+      updated_responses = build_value_updates_list
+
+      raise BulkUpdateError.new(errors), 'Bulk update error' if errors.any?
+
+      # Ensure atomic behaviour as we don't want partial inconsistent updates
+      ActiveRecord::Base.transaction do
+        apply_bulk_response_changes(updated_responses)
+        apply_person_escort_record_changes
+      end
+    end
+
+  private
+
+    def build_value_updates_list
+      [].tap do |updated_responses|
+        validator = ActiveRecord::Import::Validator.new(FrameworkResponse)
+
+        FrameworkResponse.includes(framework_question: %i[framework_flags]).where(person_escort_record: person_escort_record).find(response_values_hash.keys).each do |response|
+          response.value = response_values_hash[response.id]
+          if validator.valid_model?(response)
+            updated_responses << response if response.changed?
+          else
+            errors[response.id] = response.errors.full_messages.first
+          end
+        rescue FrameworkResponse::ValueTypeError => e
+          errors[response.id] = "Value: #{e.message} is incorrect type"
+        end
+      end
+    end
+
+    def apply_bulk_response_changes(updated_responses)
+      # Bulk update all modified response values
```

Comment:
We should return early here is `updated_responses` is empty to avoid doing any updates on unchanged responses otherwise

Reply:
Good point, also the PER update can be short circuited too (as nothing has changed), so I'll check for this scenario in the `call` method.